File: backend/app/services/telegram_bot.py
"""
Telegram Bot integration for SimNexus.
- Pushes inbound SMS to a configured Telegram chat.
- Handles /send <number> <content> and /list commands via long-polling.
"""
import asyncio
import logging
import os
from pathlib import Path
from typing import Optional

import httpx

logger = logging.getLogger(__name__)


def _load_dotenv():
    """Load .env from the app working directory into os.environ."""
    for candidate in [Path("/app/.env"), Path(".env")]:
        if candidate.exists():
            for line in candidate.read_text().splitlines():
                line = line.strip()
                if line and not line.startswith("#") and "=" in line:
                    k, v = line.split("=", 1)
                    os.environ.setdefault(k.strip(), v.strip())
            break


_load_dotenv()

BOT_TOKEN: str = os.environ.get("TELEGRAM_BOT_TOKEN", "")
CHAT_ID: str = os.environ.get("TELEGRAM_CHAT_ID", "")
API_BASE = "https://api.telegram.org"

_client: Optional[httpx.AsyncClient] = None
_poll_task: Optional[asyncio.Task] = None
_last_update_id: int = 0


def _base_url() -> str:
    return f"{API_BASE}/bot{BOT_TOKEN}"


async def _get_client() -> httpx.AsyncClient:
    global _client
    if _client is None or _client.is_closed:
        _client = httpx.AsyncClient(timeout=35.0)
    return _client


async def send_message(text: str, chat_id: Optional[str] = None, log: bool = True) -> bool:
    """Send a text message to the configured Telegram chat."""
    if not BOT_TOKEN or not (chat_id or CHAT_ID):
        return False
    target = chat_id or CHAT_ID
    try:
        client = await _get_client()
        resp = await client.post(
            f"{_base_url()}/sendMessage",
            json={"chat_id": target, "text": text, "parse_mode": "HTML"},
        )
        ok = resp.json().get("ok", False)
        if ok and log:
            _log_message(target, "SimNexus", "out", text, is_command=False)
        return ok
    except Exception as e:
        logger.error(f"Telegram sendMessage failed: {e}")
        return False


async def push_inbound_sms(modem_label: str, sender: str, content: str) -> bool:
    """Push a received SMS notification to Telegram."""
    text = (
        f"📨 <b>收到短信</b>\n"
        f"设备：{modem_label}\n"
        f"发件人：{sender}\n"
        f"内容：{content}"
    )
    return await send_message(text)


def _modem_label(modem) -> str:
    return modem.alias or modem.model or f"设备#{modem.id}"


async def _do_send(modem, number: str, content: str, chat_id: str) -> None:
    from app.services import modem_manager
    import re
    obj_path = modem.mm_object_path or ""
    if obj_path.startswith("zte:"):
        from app.services import zte_http_modem as _zte
        success = _zte.send_sms(number, content)
        err = ""
    else:
        m = re.search(r"/Modem/(\d+)$", obj_path)
        if not m:
            await send_message("❌ 无法获取设备索引", chat_id=chat_id)
            return
        success, err = modem_manager.send_sms(m.group(1), number, content)
    label = _modem_label(modem)
    if success:
        await send_message(f"✅ [{label}] 短信已发送至 {number}", chat_id=chat_id)
    else:
        await send_message(f"❌ [{label}] 发送失败：{err}", chat_id=chat_id)


def _log_message(chat_id: str, username: str, direction: str, text: str,
                 is_command: bool = False, file_id: str = None, file_type: str = None):
    """Persist a Telegram message to DB (best-effort, non-blocking)."""
    try:
        from app.core.database import SessionLocal
        from app.models.telegram import TelegramMessage
        db = SessionLocal()
        try:
            db.add(TelegramMessage(
                chat_id=chat_id, username=username,
                direction=direction, text=text, is_command=is_command,
                file_id=file_id, file_type=file_type,
            ))
            db.commit()
        finally:
            db.close()
    except Exception as e:
        logger.warning(f"Failed to log Telegram message: {e}")


async def _handle_command(message: dict) -> None:
    """Process a Telegram bot command."""
    from app.core.database import SessionLocal
    from app.models.modem import Modem, ModemStatus
    from app.models.sms import SmsMessage, SmsDirection

    chat_id = str(message["chat"]["id"])
    text: str = message.get("text") or message.get("caption") or ""
    text = text.strip()
    username = message.get("from", {}).get("username") or message.get("from", {}).get("first_name", "")

    # Handle media messages
    file_id = None
    file_type = None
    if "photo" in message:
        file_type = "photo"
        file_id = message["photo"][-1]["file_id"]  # largest size
        _log_message(chat_id, username, "in", text or "[图片]", file_id=file_id, file_type=file_type)
        return
    elif "document" in message:
        file_type = "document"
        file_id = message["document"]["file_id"]
        fname = message["document"].get("file_name", "文件")
        _log_message(chat_id, username, "in", text or f"[文件: {fname}]", file_id=file_id, file_type=file_type)
        return
    elif "video" in message:
        file_type = "video"
        file_id = message["video"]["file_id"]
        _log_message(chat_id, username, "in", text or "[视频]", file_id=file_id, file_type=file_type)
        return
    elif "sticker" in message:
        file_type = "sticker"
        file_id = message["sticker"]["file_id"]
        _log_message(chat_id, username, "in", "[贴纸]", file_id=file_id, file_type=file_type)
        return
    elif "voice" in message:
        file_type = "voice"
        file_id = message["voice"]["file_id"]
        _log_message(chat_id, username, "in", "[语音]", file_id=file_id, file_type=file_type)
        return

    if not text:
        return
    _log_message(chat_id, username, "in", text, is_command=text.startswith("/"))

    if text.startswith("/modems"):
        db = SessionLocal()
        try:
            modems = db.query(Modem).filter(Modem.is_active == True).all()
            if not modems:
                await send_message("暂无设备", chat_id=chat_id)
                return
            lines = ["📱 <b>当前设备列表</b>"]
            for m in modems:
                status = "🟢" if m.status == ModemStatus.CONNECTED else "🔴"
                label = _modem_label(m)
                phone = f" {m.phone_number}" if m.phone_number else ""
                op = f" [{m.operator}]" if m.operator else ""
                lines.append(f"{status} <b>#{m.id}</b> {label}{phone}{op}")
            lines.append("\n发送时用: /send #&lt;设备ID&gt; &lt;号码&gt; &lt;内容&gt;")
            await send_message("\n".join(lines), chat_id=chat_id)
        finally:
            db.close()

    elif text.startswith("/send"):
        # /send [#modem_id] <number> <content>
        args = text[5:].strip()
        db = SessionLocal()
        try:
            modem = None
            if args.startswith("#"):
                parts = args.split(" ", 2)
                try:
                    modem_id = int(parts[0][1:])
                    modem = db.query(Modem).filter(Modem.id == modem_id).first()
                    if not modem:
                        await send_message(f"❌ 未找到设备 #{modem_id}", chat_id=chat_id)
                        return
                    args = " ".join(parts[1:])
                except ValueError:
                    pass

            parts = args.split(" ", 1)
            if len(parts) < 2:
                await send_message(
                    "用法:\n/send &lt;号码&gt; &lt;内容&gt;\n/send #&lt;设备ID&gt; &lt;号码&gt; &lt;内容&gt;",
                    chat_id=chat_id,
                )
                return
            number, content = parts[0], parts[1]

            if modem is None:
                connected = db.query(Modem).filter(Modem.status == ModemStatus.CONNECTED).all()
                if not connected:
                    await send_message("❌ 无可用设备", chat_id=chat_id)
                    return
                if len(connected) > 1:
                    lines = ["⚠️ 有多个在线设备，请指定设备ID："]
                    for m in connected:
                        lines.append(f"  #{m.id} {_modem_label(m)}")
                    lines.append(f"\n例: /send #{connected[0].id} {number} {content}")
                    await send_message("\n".join(lines), chat_id=chat_id)
                    return
                modem = connected[0]

            await _do_send(modem, number, content, chat_id)
        finally:
            db.close()

    elif text.startswith("/list"):
        db = SessionLocal()
        try:
            # optional: /list #<modem_id>
            args = text[5:].strip()
            q = db.query(SmsMessage).filter(SmsMessage.direction == SmsDirection.INBOUND)
            if args.startswith("#"):
                try:
                    mid = int(args[1:].split()[0])
                    q = q.filter(SmsMessage.modem_id == mid)
                except ValueError:
                    pass
            msgs = q.order_by(SmsMessage.created_at.desc()).limit(10).all()
            if not msgs:
                await send_message("暂无收到的短信", chat_id=chat_id)
                return
            # group by modem
            modem_map = {m.id: _modem_label(m) for m in db.query(Modem).all()}
            lines = ["📋 <b>最近收到的短信</b>"]
            for m in msgs:
                ts = m.created_at.strftime("%m-%d %H:%M") if m.created_at else ""
                dev = modem_map.get(m.modem_id, f"#{m.modem_id}")
                lines.append(f"\n[{ts}] <b>{m.phone_number}</b> via {dev}\n{m.content}")
            await send_message("\n".join(lines), chat_id=chat_id)
        finally:
            db.close()

    elif text.startswith("/start") or text.startswith("/help"):
        help_text = (
            "🤖 <b>SimNexus Bot</b>\n\n"
            "/modems - 查看所有设备\n"
            "/list - 查看最近10条收到的短信\n"
            "/list #&lt;设备ID&gt; - 查看指定设备的短信\n"
            "/send &lt;号码&gt; &lt;内容&gt; - 发送短信（单卡时自动选择）\n"
            "/send #&lt;设备ID&gt; &lt;号码&gt; &lt;内容&gt; - 通过指定设备发送\n"
        )
        await send_message(help_text, chat_id=chat_id)


async def _poll_loop() -> None:
    """Long-poll Telegram for updates and dispatch commands."""
    global _last_update_id
    if not BOT_TOKEN:
        logger.info("Telegram bot token not configured; skipping polling")
        return
    logger.info("Telegram bot polling started")
    while True:
        try:
            client = await _get_client()
            resp = await client.get(
                f"{_base_url()}/getUpdates",
                params={"offset": _last_update_id + 1, "timeout": 30},
                timeout=35.0,
            )
            data = resp.json()
            if data.get("ok"):
                for update in data.get("result", []):
                    _last_update_id = update["update_id"]
                    msg = update.get("message") or update.get("edited_message")
                    if msg:
                        asyncio.create_task(_handle_command(msg))
        except asyncio.CancelledError:
            break
        except Exception as e:
            logger.warning(f"Telegram poll error: {e}")
            await asyncio.sleep(5)


async def start_polling() -> None:
    global _poll_task
    if not BOT_TOKEN:
        return
    _poll_task = asyncio.create_task(_poll_loop())


async def stop_polling() -> None:
    global _poll_task, _client
    if _poll_task:
        _poll_task.cancel()
        try:
            await _poll_task
        except asyncio.CancelledError:
            pass
        _poll_task = None
    if _client and not _client.is_closed:
        await _client.aclose()
        _client = None

File: backend-go/services/telegram_bot.go
package services

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"net/url"
	"regexp"
	"strconv"
	"strings"
	"time"

	"simnexus-go/config"
	"simnexus-go/database"
	"simnexus-go/models"
)

const telegramAPIBase = "https://api.telegram.org"

var (
	tgLastUpdateID int64
	tgReModemArg   = regexp.MustCompile(`^#(\d+)`)
)

func tgToken() string  { return config.C.TelegramBotToken }
func tgChatID() string { return config.C.TelegramChatID }
func tgBaseURL() string { return fmt.Sprintf("%s/bot%s", telegramAPIBase, tgToken()) }

// TelegramSendMessage sends a text message to a chat. Returns success.
func TelegramSendMessage(text, chatID string, logIt bool) bool {
	if tgToken() == "" || (chatID == "" && tgChatID() == "") {
		return false
	}
	target := chatID
	if target == "" {
		target = tgChatID()
	}
	payload, _ := json.Marshal(map[string]string{
		"chat_id": target, "text": text, "parse_mode": "HTML",
	})
	resp, err := http.Post(tgBaseURL()+"/sendMessage", "application/json", bytes.NewReader(payload))
	if err != nil {
		slog.Error("telegram sendMessage failed", "err", err)
		return false
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)
	var r struct {
		Ok bool `json:"ok"`
	}
	json.Unmarshal(body, &r)
	if r.Ok && logIt {
		tgLog(target, "SimNexus", "out", text, false, "", "")
	}
	return r.Ok
}

// TelegramPushInboundSMS notifies the Telegram chat of a received SMS.
func TelegramPushInboundSMS(modemLabel, sender, content string) bool {
	text := fmt.Sprintf("📨 <b>收到短信</b>\n设备：%s\n发件人：%s\n内容：%s", modemLabel, sender, content)
	return TelegramSendMessage(text, "", true)
}

func tgLog(chatID, username, direction, text string, isCmd bool, fileID, fileType string) {
	var up *string
	if username != "" {
		up = &username
	}
	var fid, ft *string
	if fileID != "" {
		fid = &fileID
	}
	if fileType != "" {
		ft = &fileType
	}
	database.DB.Create(&models.TelegramMessage{
		ChatID: chatID, Username: up, Direction: direction, Text: text,
		IsCommand: isCmd, FileID: fid, FileType: ft, CreatedAt: time.Now(),
	})
}

func modemLabel(m *models.Modem) string {
	if m.Alias != "" {
		return m.Alias
	}
	if m.Model != "" {
		return m.Model
	}
	return fmt.Sprintf("设备#%d", m.ID)
}

func tgDoSend(m *models.Modem, number, content, chatID string) {
	obj := m.MmObjectPath
	var success bool
	var errMsg string
	if strings.HasPrefix(obj, "zte:") {
		success = ZteSendSMS(number, content)
	} else {
		mm := reModemIdx.FindStringSubmatch(obj)
		if mm == nil {
			TelegramSendMessage("❌ 无法获取设备索引", chatID, true)
			return
		}
		success, errMsg = SendSMS(mm[1], number, content)
	}
	label := modemLabel(m)
	if success {
		TelegramSendMessage(fmt.Sprintf("✅ [%s] 短信已发送至 %s", label, number), chatID, true)
	} else {
		TelegramSendMessage(fmt.Sprintf("❌ [%s] 发送失败：%s", label, errMsg), chatID, true)
	}
}

type tgUpdate struct {
	UpdateID      int64          `json:"update_id"`
	Message       *tgMessage     `json:"message"`
	EditedMessage *tgMessage     `json:"edited_message"`
}

type tgMessage struct {
	Text    string                 `json:"text"`
	Caption string                 `json:"caption"`
	Chat    struct{ ID int64 }     `json:"chat"`
	From    map[string]interface{} `json:"from"`
	Photo   []struct {
		FileID string `json:"file_id"`
	} `json:"photo"`
	Document map[string]interface{} `json:"document"`
	Video    map[string]interface{} `json:"video"`
	Sticker  map[string]interface{} `json:"sticker"`
	Voice    map[string]interface{} `json:"voice"`
}

func tgUsername(from map[string]interface{}) string {
	if from == nil {
		return ""
	}
	if u, ok := from["username"].(string); ok && u != "" {
		return u
	}
	if u, ok := from["first_name"].(string); ok {
		return u
	}
	return ""
}

func tgHandle(msg *tgMessage) {
	chatID := strconv.FormatInt(msg.Chat.ID, 10)
	text := strings.TrimSpace(firstNonEmpty(msg.Text, msg.Caption))
	username := tgUsername(msg.From)

	// media handling
	if len(msg.Photo) > 0 {
		tgLog(chatID, username, "in", orDefault(text, "[图片]"), false, msg.Photo[len(msg.Photo)-1].FileID, "photo")
		return
	}
	if fid := mapFileID(msg.Document); fid != "" {
		fname, _ := msg.Document["file_name"].(string)
		tgLog(chatID, username, "in", orDefault(text, "[文件: "+fname+"]"), false, fid, "document")
		return
	}
	if fid := mapFileID(msg.Video); fid != "" {
		tgLog(chatID, username, "in", orDefault(text, "[视频]"), false, fid, "video")
		return
	}
	if fid := mapFileID(msg.Sticker); fid != "" {
		tgLog(chatID, username, "in", "[贴纸]", false, fid, "sticker")
		return
	}
	if fid := mapFileID(msg.Voice); fid != "" {
		tgLog(chatID, username, "in", "[语音]", false, fid, "voice")
		return
	}

	if text == "" {
		return
	}
	tgLog(chatID, username, "in", text, strings.HasPrefix(text, "/"), "", "")

	switch {
	case strings.HasPrefix(text, "/modems"):
		var modems []models.Modem
		database.DB.Where("is_active = ?", true).Find(&modems)
		if len(modems) == 0 {
			TelegramSendMessage("暂无设备", chatID, true)
			return
		}
		lines := []string{"📱 <b>当前设备列表</b>"}
		for _, m := range modems {
			st := "🔴"
			if m.Status == models.ModemConnected {
				st = "🟢"
			}
			phone := ""
			if m.PhoneNumber != "" {
				phone = " " + m.PhoneNumber
			}
			op := ""
			if m.Operator != "" {
				op = " [" + m.Operator + "]"
			}
			lines = append(lines, fmt.Sprintf("%s <b>#%d</b> %s%s%s", st, m.ID, modemLabel(&m), phone, op))
		}
		lines = append(lines, "\n发送时用: /send #&lt;设备ID&gt; &lt;号码&gt; &lt;内容&gt;")
		TelegramSendMessage(strings.Join(lines, "\n"), chatID, true)

	case strings.HasPrefix(text, "/send"):
		args := strings.TrimSpace(text[5:])
		var modem *models.Modem
		if m := tgReModemArg.FindStringSubmatch(args); m != nil {
			mid, _ := strconv.Atoi(m[1])
			var mm models.Modem
			if err := database.DB.First(&mm, mid).Error; err != nil {
				TelegramSendMessage(fmt.Sprintf("❌ 未找到设备 #%d", mid), chatID, true)
				return
			}
			modem = &mm
			args = strings.TrimSpace(args[len(m[0]):])
		}
		parts := strings.SplitN(args, " ", 2)
		if len(parts) < 2 {
			TelegramSendMessage("用法:\n/send &lt;号码&gt; &lt;内容&gt;\n/send #&lt;设备ID&gt; &lt;号码&gt; &lt;内容&gt;", chatID, true)
			return
		}
		number, content := parts[0], parts[1]
		if modem == nil {
			var connected []models.Modem
			database.DB.Where("status = ?", models.ModemConnected).Find(&connected)
			if len(connected) == 0 {
				TelegramSendMessage("❌ 无可用设备", chatID, true)
				return
			}
			if len(connected) > 1 {
				lines := []string{"⚠️ 有多个在线设备，请指定设备ID："}
				for _, m := range connected {
					lines = append(lines, fmt.Sprintf("  #%d %s", m.ID, modemLabel(&m)))
				}
				lines = append(lines, fmt.Sprintf("\n例: /send #%d %s %s", connected[0].ID, number, content))
				TelegramSendMessage(strings.Join(lines, "\n"), chatID, true)
				return
			}
			modem = &connected[0]
		}
		tgDoSend(modem, number, content, chatID)

	case strings.HasPrefix(text, "/list"):
		args := strings.TrimSpace(text[5:])
		q := database.DB.Where("direction = ?", models.SmsInbound)
		if strings.HasPrefix(args, "#") {
			if mid, err := strconv.Atoi(strings.Fields(args[1:])[0]); err == nil {
				q = q.Where("modem_id = ?", mid)
			}
		}
		var msgs []models.SmsMessage
		q.Order("created_at desc").Limit(10).Find(&msgs)
		if len(msgs) == 0 {
			TelegramSendMessage("暂无收到的短信", chatID, true)
			return
		}
		var modems []models.Modem
		database.DB.Find(&modems)
		labelMap := map[uint]string{}
		for i := range modems {
			labelMap[modems[i].ID] = modemLabel(&modems[i])
		}
		lines := []string{"📋 <b>最近收到的短信</b>"}
		for _, m := range msgs {
			ts := m.CreatedAt.Format("01-02 15:04")
			dev := labelMap[m.ModemID]
			if dev == "" {
				dev = fmt.Sprintf("#%d", m.ModemID)
			}
			lines = append(lines, fmt.Sprintf("\n[%s] <b>%s</b> via %s\n%s", ts, m.PhoneNumber, dev, m.Content))
		}
		TelegramSendMessage(strings.Join(lines, "\n"), chatID, true)

	case strings.HasPrefix(text, "/start"), strings.HasPrefix(text, "/help"):
		help := "🤖 <b>SimNexus Bot</b>\n\n" +
			"/modems - 查看所有设备\n" +
			"/list - 查看最近10条收到的短信\n" +
			"/list #&lt;设备ID&gt; - 查看指定设备的短信\n" +
			"/send &lt;号码&gt; &lt;内容&gt; - 发送短信（单卡时自动选择）\n" +
			"/send #&lt;设备ID&gt; &lt;号码&gt; &lt;内容&gt; - 通过指定设备发送\n"
		TelegramSendMessage(help, chatID, true)
	}
}

func mapFileID(m map[string]interface{}) string {
	if m == nil {
		return ""
	}
	if v, ok := m["file_id"].(string); ok {
		return v
	}
	return ""
}

func orDefault(s, def string) string {
	if s == "" {
		return def
	}
	return s
}

// StartTelegramPolling begins long-polling getUpdates until ctx is cancelled.
func StartTelegramPolling(ctx context.Context) {
	if tgToken() == "" {
		slog.Warn("Telegram bot token not configured; skipping polling")
		return
	}
	slog.Info("Telegram bot polling started")
	client := &http.Client{Timeout: 35 * time.Second}
	for {
		select {
		case <-ctx.Done():
			return
		default:
		}
		u := fmt.Sprintf("%s/getUpdates?offset=%d&timeout=30", tgBaseURL(), tgLastUpdateID+1)
		resp, err := client.Get(u)
		if err != nil {
			time.Sleep(5 * time.Second)
			continue
		}
		body, _ := io.ReadAll(resp.Body)
		resp.Body.Close()
		var data struct {
			Ok     bool       `json:"ok"`
			Result []tgUpdate `json:"result"`
		}
		if json.Unmarshal(body, &data) != nil || !data.Ok {
			continue
		}
		for _, up := range data.Result {
			tgLastUpdateID = up.UpdateID
			msg := up.Message
			if msg == nil {
				msg = up.EditedMessage
			}
			if msg != nil {
				go tgHandle(msg)
			}
		}
	}
}

// TelegramProxyFile downloads a Telegram file, returning content, content-type, filename.
func TelegramProxyFile(fileID string) ([]byte, string, string, error) {
	client := &http.Client{Timeout: 30 * time.Second}
	r, err := client.Get(fmt.Sprintf("%s/getFile?file_id=%s", tgBaseURL(), url.QueryEscape(fileID)))
	if err != nil {
		return nil, "", "", err
	}
	defer r.Body.Close()
	body, _ := io.ReadAll(r.Body)
	var data struct {
		Ok     bool `json:"ok"`
		Result struct {
			FilePath string `json:"file_path"`
		} `json:"result"`
	}
	if json.Unmarshal(body, &data) != nil || !data.Ok {
		return nil, "", "", fmt.Errorf("file not found")
	}
	fileURL := fmt.Sprintf("%s/file/bot%s/%s", telegramAPIBase, tgToken(), data.Result.FilePath)
	resp, err := client.Get(fileURL)
	if err != nil {
		return nil, "", "", err
	}
	defer resp.Body.Close()
	content, _ := io.ReadAll(resp.Body)
	ct := resp.Header.Get("Content-Type")
	if ct == "" {
		ct = "application/octet-stream"
	}
	parts := strings.Split(data.Result.FilePath, "/")
	return content, ct, parts[len(parts)-1], nil
}

// TelegramSendFile uploads a photo/document to the configured chat.
func TelegramSendFile(filename string, content []byte, contentType, caption string) (bool, string, string, string) {
	if tgToken() == "" || tgChatID() == "" {
		return false, "", "", "Bot not configured"
	}
	isImage := strings.HasPrefix(contentType, "image/")
	method := "sendDocument"
	field := "document"
	fileType := "document"
	if isImage {
		method, field, fileType = "sendPhoto", "photo", "photo"
	}

	var buf bytes.Buffer
	boundary := "----simnexusboundary"
	writePart := func(name, value string) {
		buf.WriteString("--" + boundary + "\r\n")
		buf.WriteString(fmt.Sprintf("Content-Disposition: form-data; name=%q\r\n\r\n", name))
		buf.WriteString(value + "\r\n")
	}
	writePart("chat_id", tgChatID())
	writePart("caption", caption)
	buf.WriteString("--" + boundary + "\r\n")
	buf.WriteString(fmt.Sprintf("Content-Disposition: form-data; name=%q; filename=%q\r\n", field, filename))
	buf.WriteString("Content-Type: " + contentType + "\r\n\r\n")
	buf.Write(content)
	buf.WriteString("\r\n--" + boundary + "--\r\n")

	req, _ := http.NewRequest(http.MethodPost, fmt.Sprintf("%s/%s", tgBaseURL(), method), &buf)
	req.Header.Set("Content-Type", "multipart/form-data; boundary="+boundary)
	resp, err := (&http.Client{Timeout: 60 * time.Second}).Do(req)
	if err != nil {
		return false, "", "", err.Error()
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)
	var data struct {
		Ok          bool                   `json:"ok"`
		Description string                 `json:"description"`
		Result      map[string]interface{} `json:"result"`
	}
	json.Unmarshal(body, &data)
	if !data.Ok {
		return false, "", "", data.Description
	}
	sentFileID := ""
	if isImage {
		if arr, ok := data.Result["photo"].([]interface{}); ok && len(arr) > 0 {
			if last, ok := arr[len(arr)-1].(map[string]interface{}); ok {
				sentFileID, _ = last["file_id"].(string)
			}
		}
	} else if doc, ok := data.Result["document"].(map[string]interface{}); ok {
		sentFileID, _ = doc["file_id"].(string)
	}
	return true, fileType, sentFileID, ""
}
